File: app/src/main/java/com/litroenade/yunjiweather/data/repository/CityRepository.java
package com.litroenade.yunjiweather.data.repository;

import com.litroenade.yunjiweather.data.entity.CityEntity;
import com.litroenade.yunjiweather.data.local.CityDao;
import com.litroenade.yunjiweather.utils.DefaultCityUtils;

import java.util.List;

public final class CityRepository {

    private final CityDao cityDao;

    public CityRepository(CityDao cityDao) {
        this.cityDao = cityDao;
    }

    public CityEntity resolveDefaultCity(long nowTime) {
        return DefaultCityUtils.resolveDefaultCity(cityDao, nowTime);
    }

    public List<CityEntity> findAll() {
        return cityDao.findAll();
    }

    public CityEntity findDefaultCity() {
        return cityDao.findDefaultCity();
    }

    public CityEntity findByLocationId(String locationId) {
        return cityDao.findByLocationId(locationId);
    }

    public int count() {
        return cityDao.count();
    }

    public void insert(CityEntity city) {
        cityDao.insert(city);
    }

    public void setDefaultCity(String locationId, long nowTime) {
        cityDao.clearDefaultCity();
        cityDao.setDefaultCity(locationId, nowTime);
    }

    public void deleteCity(CityEntity city, long nowTime) {
        cityDao.deleteByLocationId(city.locationId);
        if (!city.isDefault) {
            return;
        }
        List<CityEntity> remainingCities = cityDao.findAll();
        if (!remainingCities.isEmpty()) {
            setDefaultCity(remainingCities.get(0).locationId, nowTime);
        }
    }

    public void saveAsDefaultCity(CityEntity city, long nowTime) {
        CityEntity oldCity = cityDao.findByLocationId(city.locationId);
        if (oldCity == null) {
            cityDao.clearDefaultCity();
            cityDao.insert(city);
            return;
        }
        setDefaultCity(oldCity.locationId, nowTime);
    }
}

File: app/src/main/java/com/litroenade/yunjiweather/data/repository/CityRepository.kt
package com.litroenade.yunjiweather.data.repository

import com.litroenade.yunjiweather.data.entity.CityEntity
import com.litroenade.yunjiweather.data.local.CityDao
import com.litroenade.yunjiweather.utils.DefaultCityUtils

class CityRepository(private val cityDao: CityDao) {

    fun resolveDefaultCity(nowTime: Long): CityEntity {
        return DefaultCityUtils.resolveDefaultCity(cityDao, nowTime)
    }

    fun findAll(): List<CityEntity> {
        return cityDao.findAll()
    }

    fun findDefaultCity(): CityEntity? {
        return cityDao.findDefaultCity()
    }

    fun findByLocationId(locationId: String): CityEntity? {
        return cityDao.findByLocationId(locationId)
    }

    fun count(): Int {
        return cityDao.count()
    }

    fun insert(city: CityEntity) {
        cityDao.insert(city)
    }

    fun setDefaultCity(locationId: String, nowTime: Long) {
        cityDao.clearDefaultCity()
        cityDao.setDefaultCity(locationId, nowTime)
    }

    fun deleteCity(city: CityEntity, nowTime: Long) {
        cityDao.deleteByLocationId(city.locationId)
        if (!city.isDefault) {
            return
        }
        val remainingCities = cityDao.findAll()
        if (remainingCities.isNotEmpty()) {
            setDefaultCity(remainingCities[0].locationId, nowTime)
        }
    }

    fun moveCity(locationId: String, direction: Int, nowTime: Long) {
        if (direction == 0) {
            return
        }
        val cities = cityDao.findAll()
        val currentIndex = cities.indexOfFirst { city -> city.locationId == locationId }
        if (currentIndex < 0) {
            return
        }
        val city = cities[currentIndex]
        if (city.isDefault) {
            return
        }
        val targetIndex = (currentIndex + direction).coerceIn(0, cities.lastIndex)
        val targetCity = cities[targetIndex]
        if (targetIndex == currentIndex || targetCity.isDefault) {
            return
        }
        val currentSortOrder = city.sortOrder
        val targetSortOrder = targetCity.sortOrder
        cityDao.updateSortOrder(city.locationId, targetSortOrder, nowTime)
        cityDao.updateSortOrder(targetCity.locationId, currentSortOrder, nowTime)
    }

    fun saveAsDefaultCity(city: CityEntity, nowTime: Long) {
        val oldCity = cityDao.findByLocationId(city.locationId)
        if (oldCity == null) {
            cityDao.clearDefaultCity()
            cityDao.insert(city)
            return
        }
        setDefaultCity(oldCity.locationId, nowTime)
    }
}
